lib/membrane_rpicam/source.ex

```diff
@@ -57,18 +65,35 @@ defmodule Membrane.Rpicam.Source do
 
     buffer = %Buffer{payload: data, pts: time - init_time}
 
-    {[buffer: {:output, buffer}], %{state | init_time: init_time}}
+    {[buffer: {:output, buffer}], %{state | init_time: init_time, camera_open: true}}
   end
 
   @impl true
   def handle_info({port, {:exit_status, exit_status}}, _ctx, %{app_port: port} = state) do
-    if exit_status == 0 do
-      {[end_of_stream: :output], state}
-    else
-      raise "#{@app_name} error, exit status: #{exit_status}"
+    cond do
+      exit_status == 0 ->
+        {[end_of_stream: :output], state}
+
+      not state.camera_open and state.retries == @max_retries ->
+        raise "Max retries exceeded, camera failed to open, exit status: #{exit_status}"
+
+      not state.camera_open ->
+        Port.close(port)
+        Membrane.Logger.warning("Camera failed to open with exit status #{exit_status}, retrying")
+        Process.sleep(50)
+        new_port = open_port(state.options)
+        {[], %{state | retries: state.retries + 1, app_port: new_port}}
+
+      true ->
+        raise "#{@app_name} error, exit status: #{exit_status}"
```

Comment:
I think it's slightly cleaner this way

```suggestion
     state.camera_open ->
       raise "#{@app_name} error, exit status: #{exit_status}"

      state.retries < @max_retries ->
        Port.close(port)
        Membrane.Logger.warning("Camera failed to open with exit status #{exit_status}, retrying")
        Process.sleep(50)
        new_port = open_port(state.options)
        {[], %{state | retries: state.retries + 1, app_port: new_port}}
        
       true ->
         raise "Max retries exceeded, camera failed to open, exit status: #{exit_status}"
```